framework/src/base/MooseApp.C

```diff
@@ -438,3 +452,205 @@ MooseApp::getOutputWarehouse()
+  std::string dl_lib_filename;
+
+  // This RE looks for absolute path libtool filenames (i.e. begins with a slash and ends with a .la)
+  pcrecpp::RE re_deps("(/\\S*\\.la)");
+
+  std::ifstream handle(library_filename.c_str());
+  if (handle.is_open())
+  {
+    while (std::getline(handle, line))
+    {
+      // Look for the system dependent dynamic library filename to open
+      if (line.find("dlname=") != std::string::npos)
+        // Magic numbers are computed from length of this string "dlname=' and line minus that string plus quotes"
+        dl_lib_filename = line.substr(8, line.size()-9);
+
+      if (line.find("dependency_libs=") != std::string::npos)
+      {
+        pcrecpp::StringPiece input(line);
+        pcrecpp::StringPiece depend_library;
+        while (re_deps.FindAndConsume(&input, &depend_library))
+          // Recurse here to load dependent libraries in depth-first order
+          loadLibraryAndDependencies(depend_library.as_string(), registration_method_name, factory);
+
+        // There's only one line in the .la file containing the dependency libs so break after finding it
+        break;
+      }
+    }
+    handle.close();
+  }
+
+  // Time to load the library, First see if we've already loaded this particular dynamic library
+  if (_lib_handles.find(library_filename) == _lib_handles.end() && // make sure we haven't already loaded this library
+      dl_lib_filename != "")                                       // AND make sure we have a library name (we won't for static linkage)
+  {
+    std::pair<std::string, std::string> lib_name_parts = MooseUtils::splitFileName(library_filename);
+
+    // Assemble the actual filename using the base path of the *.la file and the dl_lib_filename
+    std::string dl_lib_full_path = lib_name_parts.first + '/' + dl_lib_filename;
+
+    void * handle = dlopen(dl_lib_full_path.c_str(), RTLD_LAZY);
+    if (!handle)
+      mooseError("Cannot open library:\n" << dlerror());
+
+    // Reset errors
+    dlerror();
+
+    // get the name of the unique registration method
+//    std::string reg_function = libNameToAppName(lib_name_parts.second) + "App__registerApps";
+
+    // get the pointer to the method in the library
+    void * registration_method = dlsym(handle, registration_method_name.c_str());
+
+    // Choose the function signature based on whether or not factory parameter to this method is NULL
+    if (factory)
+    {
+      typedef void (*register_app_t)(Factory *);
+      register_app_t *reg_ptr = reinterpret_cast<register_app_t *>( &registration_method );
+
+      // Catch errors
+      const char *dlsym_error = dlerror();
+      if (dlsym_error)
+      {
+#ifdef DEBUG
+        mooseWarning("Unable to find extern \"C\" method \"" << registration_method_name << "\" in library: " << dl_lib_full_path << ".\n"
+                     << "This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.\n");
+#endif
```

Comment:
This is a good idea.  We'll have to see if we get a lot of questions from people that see it...